wgpu-core/src/device/life.rs

```diff
@@ -259,25 +268,30 @@ impl<B: hal::Backend> LifetimeTracker<B> {
         });
     }
 
-    fn wait_idle(&self, device: &B::Device) {
+    fn wait_idle(&self, device: &B::Device) -> Result<(), WaitIdleError> {
         if !self.active.is_empty() {
             log::debug!("Waiting for IDLE...");
             let status = unsafe {
                 device.wait_for_fences(
                     self.active.iter().map(|a| &a.fence),
                     hal::device::WaitFor::All,
                     CLEANUP_WAIT_MS * 1_000_000,
-                )
+                )?
             };
             log::debug!("...Done");
-            assert_eq!(status, Ok(true), "GPU got stuck :(");
+
+            if status == false {
+                // We timed out while waiting for the fences
+                return Err(WaitIdleError::StuckGpu);
+            }
         }
+        Ok(())
     }
 
     /// Returns the last submission index that is done.
     pub fn triage_submissions(&mut self, device: &B::Device, force_wait: bool) -> SubmissionIndex {
         if force_wait {
-            self.wait_idle(device);
+            self.wait_idle(device).unwrap();
```

Comment:
I've looked into pulling this error out and returning it, but it's quite deep in the API. I don't know if it's desirable to do so.

Reply:
I think we'd need that, actually. Looking at the code that calls it
- `maintain` should return `Result`
- `prepare_to_die` should just log an error and early return, it doesn't need to unwrap or return a result
- `device_wait_for_buffer` should return `Result`

Reply:
Alright, I've made the required changes to get this error returned upwards.